pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,266 @@
+	common.Service
+	SubnetStore *SubnetStore
+}
+
+// SubnetParameters stores parameters to CRUD Subnet object
+type SubnetParameters struct {
+	OrgID     string
+	ProjectID string
+	VPCID     string
+}
+
+var subnetService *SubnetService
+var lock = &sync.Mutex{}
+
+// GetSubnetService get singleton SubnetService instance, subnet/subnetset controller share the same instance.
+func GetSubnetService(service common.Service) *SubnetService {
+	if subnetService == nil {
+		lock.Lock()
+		defer lock.Unlock()
+		if subnetService == nil {
+			var err error
+			if subnetService, err = InitializeSubnetService(service); err != nil {
+				log.Error(err, "failed to initialize subnet commonService")
+				os.Exit(1)
+			}
+		}
+	}
+	return subnetService
+}
+
+// InitializeSubnetService initialize Subnet service.
+func InitializeSubnetService(service common.Service) (*SubnetService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+	subnetService := &SubnetService{
+		Service: service,
+		SubnetStore: &SubnetStore{
+			ResourceStore: common.ResourceStore{
+				Indexer: cache.NewIndexer(keyFunc, cache.Indexers{
+					common.TagScopeSubnetCRUID: subnetIndexFunc,
+				}),
+				BindingType: model.VpcSubnetBindingType(),
+			},
+		},
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, nil, subnetService.SubnetStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetService, err
+	}
+	return subnetService, nil
+}
+
+// TODO Test update of VpcSubnet(eg. update tags)
+func (service *SubnetService) CreateOrUpdateSubnet(obj client.Object, vpcInfo *common.VPCResourceInfo, tags []model.Tag) (string, error) {
+	uid := string(obj.GetUID())
+	nsxSubnet, err := service.buildSubnet(obj, tags)
+	if err != nil {
+		log.Error(err, "failed to build Subnet")
+		return "", err
+	}
+	// Only check whether needs update when obj is v1alpha1.Subnet
+	if _, ok := obj.(*v1alpha1.Subnet); ok {
+		existingSubnet := service.SubnetStore.GetByKey(uid)
+		changed := false
+		if existingSubnet == nil {
+			changed = true
+		} else {
+			changed = common.CompareResource(SubnetToComparable(existingSubnet), SubnetToComparable(nsxSubnet))
+		}
+		if !changed {
+			log.Info("subnet not changed, skip updating", "subnet.Id", uid)
+			return uid, nil
+		}
+	}
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, vpcInfo)
+	if err != nil {
+		log.Error(err, "WrapHierarchySubnet failed")
+		return "", err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return "", err
+	}
+	// Get Subnet from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	if *nsxSubnet, err = service.NSXClient.SubnetsClient.Get(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, *nsxSubnet.Id); err != nil {
+		return "", err
+	}
```

Comment:
Please log the error before returning it. Because this function `CreateOrUpdateSubnet` may be indirectly invoked by `GetAvailableSubnet`. Imagine that when we got a standalone error `"com.vmware.vapi.std.errors.invalid_request"` from `GetAvailableSubnet`, locating the issue will be hard.

Reply:
Make sense, done.